File: src/commands/snapshot.ts
import fs from "node:fs";
import path from "node:path";
import { captureSnapshot } from "../runtime/snapshot.js";
import { encodeToken } from "../runtime/token.js";
import { EXIT_OK } from "../core/exit-codes.js";
import { rule, ui } from "../utils/logger.js";

export interface SnapshotOptions {
  rootDir: string;
  output?: string;
  token: boolean;
  json: boolean;
  globals: boolean;
}

/** `envdoctor snapshot` — capture this machine's live runtime. */
export async function runSnapshot(opts: SnapshotOptions): Promise<number> {
  const snapshot = await captureSnapshot({ globals: opts.globals });

  if (opts.output) {
    const dest = path.resolve(opts.rootDir, opts.output);
    fs.writeFileSync(dest, JSON.stringify(snapshot, null, 2) + "\n");
    process.stderr.write(`${ui.success("✓")} Snapshot written to ${ui.location(opts.output)}\n`);
  }

  if (opts.json) {
    process.stdout.write(JSON.stringify(snapshot, null, 2) + "\n");
    return EXIT_OK;
  }

  if (opts.token) {
    process.stdout.write(encodeToken(snapshot) + "\n");
    return EXIT_OK;
  }

  // Human summary.
  const title = "RUNTIME SNAPSHOT";
  process.stdout.write(ui.title(title) + "\n");
  process.stdout.write(`${rule(title.length * 2)}\n\n`);
  process.stdout.write(`  ${ui.bold("OS")}  ${snapshot.os.platform}/${snapshot.os.arch} ${ui.dim(snapshot.os.release)}\n\n`);

  process.stdout.write(`  ${ui.section("Tools")}\n`);
  if (snapshot.tools.length === 0) {
    process.stdout.write(`  ${ui.dim("none detected")}\n`);
  } else {
    for (const t of snapshot.tools) {
      process.stdout.write(`  ${ui.same()} ${ui.name(t.tool.padEnd(8))} ${t.version}  ${ui.dim(t.resolvedFrom)}\n`);
    }
  }

  process.stdout.write(`\n  ${ui.section(`PATH (${snapshot.path.length} entries)`)}\n`);
  snapshot.path.slice(0, 12).forEach((p, i) => {
    process.stdout.write(`  ${ui.dim(String(i + 1).padStart(2))}  ${p}\n`);
  });
  if (snapshot.path.length > 12) {
    process.stdout.write(`  ${ui.dim(`… ${snapshot.path.length - 12} more`)}\n`);
  }

  const ecosystems = Object.keys(snapshot.globals);
  if (ecosystems.length > 0) {
    process.stdout.write(`\n  ${ui.section("Globals")}\n`);
    for (const eco of ecosystems) {
      process.stdout.write(`  ${ui.dim(eco)}: ${(snapshot.globals[eco] ?? []).length} packages\n`);
    }
  } else if (!opts.globals) {
    process.stdout.write(`\n  ${ui.dim("Globals omitted — pass --globals to include the package inventory.")}\n`);
  }

  process.stdout.write(
    `\n  ${ui.dim(`Share with:  envdoctor snapshot --token   ·   compare with:  envdoctor snapshot-diff <a> <b>`)}\n`,
  );

  return EXIT_OK;
}

File: dart/lib/src/commands/snapshot.dart
/// `envdoctor snapshot` — capture this machine's live runtime.
library;

import 'dart:convert';
import 'dart:io';

import '../core/exit_codes.dart';
import '../models/runtime_snapshot.dart';
import '../runtime/capture.dart';
import '../runtime/token.dart';
import '../utils/json.dart';
import '../utils/paths.dart';
import 'shared.dart';

class SnapshotOptions {
  final String rootDir;
  final String? output;
  final bool token;
  final bool json;
  final bool globals;

  SnapshotOptions({
    required this.rootDir,
    this.output,
    this.token = false,
    this.json = false,
    this.globals = false,
  });
}

int runSnapshot(SnapshotOptions opts) {
  final snapshot = captureSnapshot(globals: opts.globals);

  if (opts.output != null) {
    final dest = _resolvePath(opts.rootDir, opts.output!);
    File(dest).writeAsStringSync('${_snapshotJsonPretty(snapshot)}\n');
    err('✓ Snapshot written to ${opts.output}\n');
  }

  if (opts.json) {
    out('${_snapshotJsonPretty(snapshot)}\n');
    return ExitCodes.exitOk;
  }

  if (opts.token) {
    out('${encodeToken(snapshot)}\n');
    return ExitCodes.exitOk;
  }

  // Human summary.
  const title = 'RUNTIME SNAPSHOT';
  out('$title\n');
  out('${'─' * (title.length * 2)}\n\n');
  out('  OS  ${snapshot.os.platform}/${snapshot.os.arch} ${snapshot.os.release}\n\n');

  out('  Tools\n');
  if (snapshot.tools.isEmpty) {
    out('  none detected\n');
  } else {
    for (final t in snapshot.tools) {
      out('  ✓ ${t.tool.padRight(8)} ${t.version}  ${t.resolvedFrom}\n');
    }
  }

  out('\n  PATH (${snapshot.path.length} entries)\n');
  final shown = snapshot.path.take(12).toList();
  for (var i = 0; i < shown.length; i++) {
    out('  ${(i + 1).toString().padLeft(2)}  ${shown[i]}\n');
  }
  if (snapshot.path.length > 12) {
    out('  … ${snapshot.path.length - 12} more\n');
  }

  final ecosystems = snapshot.globals.keys.toList();
  if (ecosystems.isNotEmpty) {
    out('\n  Globals\n');
    for (final eco in ecosystems) {
      out('  $eco: ${snapshot.globals[eco]!.length} packages\n');
    }
  } else if (!opts.globals) {
    out('\n  Globals omitted — pass --globals to include the package inventory.\n');
  }

  out(
      '\n  Share with:  envdoctor snapshot --token   ·   compare with:  envdoctor snapshot-diff <a> <b>\n');

  return ExitCodes.exitOk;
}

String _resolvePath(String rootDir, String p) =>
    isAbsolutePath(p) ? p : joinPath(rootDir, p);

/// Pretty-print the same shape the token codec emits compactly.
String _snapshotJsonPretty(RuntimeSnapshot snapshot) {
  final compact = snapshotToJson(snapshot);
  return jsonPretty(_toPlain(jsonDecode(compact)));
}

Object? _toPlain(Object? el) => switch (el) {
      Map<String, Object?> map => JsonObject()
        ..entries.addAll(map.entries.map((e) => (e.key, _toPlain(e.value)))),
      Map map => JsonObject()
        ..entries.addAll(map.entries.map((e) => (e.key.toString(), _toPlain(e.value)))),
      List list => list.map(_toPlain).toList(),
      _ => el,
    };
